src/transformers/models/jamba/modeling_jamba.py

```diff
@@ -0,0 +1,2127 @@
+    # Ignore copy
+    def forward(
+        self,
+        input_ids: torch.LongTensor = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_values: Optional[List[torch.FloatTensor]] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        labels: Optional[torch.LongTensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        output_router_logits: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        calc_logits_for_entire_prompt: Optional[bool] = True,
+    ) -> Union[Tuple, MoeCausalLMOutputWithPast]:
+        r"""
+        Args:
+            labels (`torch.LongTensor` of shape `(batch_size, sequence_length)`, *optional*):
+                Labels for computing the masked language modeling loss. Indices should either be in `[0, ...,
+                config.vocab_size]` or -100 (see `input_ids` docstring). Tokens with indices set to `-100` are ignored
+                (masked), the loss is only computed for the tokens with labels in `[0, ..., config.vocab_size]`.
+
+            calc_logits_for_entire_prompt (`bool`, *optional*):
+                Whether or not to calculate the logits for the entire prompt, or just the last token. Only last token
+                logits are needed for generation, and calculating them only for that token can save memory,
+                which becomes pretty significant for long sequences.
+
+        Returns:
+        ```"""
+
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_router_logits = (
+            output_router_logits if output_router_logits is not None else self.config.output_router_logits
+        )
+
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        # decoder outputs consists of (dec_features, layer_state, dec_hidden, dec_attn)
+        outputs = self.model(
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            position_ids=position_ids,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            output_router_logits=output_router_logits,
+            return_dict=return_dict,
+        )
+
+        hidden_states = outputs[0]
+        if calc_logits_for_entire_prompt:
+            logits = self.lm_head(hidden_states)
+        else:
+            logits = self.lm_head(hidden_states[..., -1:, :])
```

Comment:
Mmmm could you explain the motivations behind this?

Reply:
Sure
it's pretty much explained in the docstring for `calc_logits_for_entire_prompt` in `configuration_jamba.py`. For long sequences, the logits can take a lot of GPU memory, especially as they are saved in FP32. So for a prompt of 128K tokens, with our vocab size of 64K, only the logits for the prompt take 32GB of GPU memory (128K*64K*4). The thing is that in order to generate from the model, we don't need all the prompt logits - just those of the last token. Anyway the GenerationMixin takes only the logits of the last prompt token (`next_token_logits = outputs.logits[:, -1, :]` appears many times in `src/transformers/generation/utils.py`). So we want to save all this unnecessary memory and compute only the logits we need. 
Honestly, we were a bit surprised to see that the standard in `transformers` is to calculate the logits for the entire prompt when generating. I understand that for relatively short prompts this doesn't add up to a lot of extra memory, but for long prompt it's a complete waste.

Reply:
That's where I say, when we generate with transformers, we only pass a single input ids starting from the second forward pass. Which is why we never need this, the logits generated after the first forward pass are always of shape 1 in sequence length! 
This can be safely removed

Reply:
Ah sorry you mean the first forward as well. Sure let’s keep it

Reply:
Yeah I was talking about the first forward pass.
RE assisted generation - you're right and that's why we kept this as a config option. If a user wants to use assisted generation, they can set `calc_logits_for_entire_prompt` to True in the config and everything will work.
As you saw, part of the Jamba promise is to be able to fit long sequences (~140K) on a single 80GB gpu (with int8 weight quantization). If we calculate the logits for the entire propmt, that's not possible. That's the reason we feel that by default entire prompt logits shouldn't be calculated during generation. If the user wants/needs that, they can do that by setting the appropriate attribute in the config

Reply:
Alright. Let's leave it for now, this will be streamlined to `generate` as this can be a generation config argument set if you use `use_cache` and no `assisted_decoding`. 
Will be deprecated in a near futur, as this is mostly for inference

Reply:
@gante - just to make sure I understand: If we do that, we'll still need to modify the assisted generation code to set `num_logits_to_keep` to `candidate_length` correct?